### PR TITLE
Uditkumar01/interview/add pick slot feat

### PR DESCRIPTION
**Pull request by uditkumar01**
# 🔥 Interview System 🔥 

### Roles in use:
- Interviewer
- AC Interviewer
- User

### Steps:
- Opening interview step of N users (Admin)
- Fill admission form (User)
- Review admission form (Interviewer/AC Interviewer)
- Pay for interview (with razorpay intergration for non FA Users)
- Open Slots (Interviewer [for non FA Users]/ AC Interviewer [for FA Users])
- Schedule Interview Slot (User)
- GMeet Intergration for automating meet creation process
- Take Interviewer and give mark to student (Interviewer/AC Interviewer)
- Result released (one of  passed/rejected/parked)

### CODE REVIEW

File: client/src/components/AlertDialogBox/AlertDialogBox.tsx [hunk 1]
```diff
@@ -17,12 +17,18 @@ export function AlertDialogBox({
...
+  rejectButtonText,
 }: {
   btntext: string
   title: string
+  rejectButtonText?: string
+  acceptButtonText?: string
   description?: string
   btnProps?: any
   callback?: any
+  onCloseCallback?: any
```
Comment (SAYUK09) [inline client/src/components/AlertDialogBox/AlertDialogBox.tsx:31]: Why do we `any` type here?

File: client/src/components/InterviewSlot/InterviewSlot.tsx [hunk 2]
```diff
@@ -89,28 +105,30 @@ const SingleSlot = ({
...
-            </>
+              {/* <AlertDialogBox
+                btntext="Cancel Interview"
+                title="Cancel Interview"
+                btnProps={{
+                  variant: 'ghost',
+                }}
+                description="Are you sure you want to cancel your interview?"
+                callback={cancelInterviewHandler}
+              /> */}
```
Comment (SAYUK09) [inline client/src/components/InterviewSlot/InterviewSlot.tsx:130]: Let's remove the comments if not necessary

File: client/src/context/PaymentContext/PaymentContext.tsx [hunk 3]
```diff
@@ -63,9 +64,22 @@ export const PaymentDetailsProvider = ({
               status: 'success',
               setLoading,
             }
-            const payRes = await paymentHandler(paymentHandlerArguments)
-            console.log(payRes)
+            const payRes: any = await paymentHandler(paymentHandlerArguments)
```
Comment (SAYUK09) [inline client/src/context/PaymentContext/PaymentContext.tsx:67]: Can we give this an appropriate type?

File: client/src/context/PaymentContext/PaymentContext.tsx [hunk 4]
```diff
@@ -63,9 +64,22 @@ export const PaymentDetailsProvider = ({
...
+              //   user: {
+              //     ...(authState?.user || {}),
+              //     ...(payRes?.user || {}),
+              //   },
+              // }))
+
+              // reload page after 3 second
+              setTimeout(() => {
+                router.reload()
+              }, 3000)
```
Comment (SAYUK09) [inline client/src/context/PaymentContext/PaymentContext.tsx:81]: Curious, why are we reloading the page after 3sec?

**Review by SAYUK09 — CHANGES_REQUESTED** [review]
This is great work! Here are few suggestions.

File: client/src/components/AlertDialogBox/AlertDialogBox.tsx (hunk 1 above)
Comment (uditkumar01) [reply client/src/components/AlertDialogBox/AlertDialogBox.tsx:31]: cause its a reusable components and this callback could be async, non async, with params, without params. So, its better to key it as any.
Please let me know if there is a better alternative.

File: client/src/components/InterviewSlot/InterviewSlot.tsx (hunk 2 above)
Comment (uditkumar01) [reply client/src/components/InterviewSlot/InterviewSlot.tsx:130]: sure

File: client/src/context/PaymentContext/PaymentContext.tsx (hunk 3 above)
Comment (uditkumar01) [reply client/src/context/PaymentContext/PaymentContext.tsx:67]: yes

File: client/src/context/PaymentContext/PaymentContext.tsx (hunk 4 above)
Comment (uditkumar01) [reply client/src/context/PaymentContext/PaymentContext.tsx:81]: there was some issue while updating the state so this was just a temporary fix.
Will fix it soon

**Review by sohamsshah — CHANGES_REQUESTED** [review]
Kudos team! Such a well-written piece💯 Keep it up!
I have made few comments related to enhancements. You can take a look at them. Else LGTM!

File: client/src/components/StepCard/StepCardInteview.tsx [hunk 5]
```diff
@@ -24,9 +24,16 @@ export function StepCardInterview({
   let interviewSubmissionstatus = ''
   if (status.level == step.level) {
     interviewSubmissionstatus = 'Not Started'
-    if (status.status == 'admission_form_incomplete') {
+    if (status.status == 'admission_form_pending') {
       interviewSubmissionstatus = 'Not Started'
-    } else if (status.status == 'admission_form_under_review') {
+    } else if (
+      status.status == 'admission_form_under_review' ||
```
[the remark this answers is not in the document]
Comment (uditkumar01) [reply client/src/components/StepCard/StepCardInteview.tsx:30]: yes 👍

File: client/src/pages/interviewer/add-slot.tsx [hunk 6]
```diff
@@ -29,53 +54,119 @@ function AddSlot() {
...
+      currentDate.setFullYear(selectedDate.year)
       const prevDate = currentDate.toISOString()
       const timeAfter30Mins = addTime(currentDate, 30)
-      const payload = {
-        from: prevDate,
-        to: timeAfter30Mins.toISOString(),
-      }
-      const res = await addTimeSlot(payload)
+      const isAnyIntervalOverlapping = interviewerState?.slots?.some((slot) => {
+        const from1 = new Date(prevDate)
```
[the remark this answers is not in the document]
Comment (uditkumar01) [reply client/src/pages/interviewer/add-slot.tsx:116]: sure 👍

File: src/controllers/InterviewController.ts [hunk 7]
```diff
@@ -243,3 +245,257 @@ export const deleteInterviewerTimeSlot: RequestHandler = async (
...
+        return res.status(400).json({
+          success: false,
+          message: 'You have reached the maximum reschedule limit',
+        })
+      }
+
+      const updatedSlot = new Slot(
+        extend(slot, {
+          status: 'open',
+          link: undefined,
```
[the remark this answers is not in the document]
Comment (uditkumar01) [reply src/controllers/InterviewController.ts:297]: 👍

File: src/controllers/InterviewController.ts [hunk 8]
```diff
@@ -243,3 +245,257 @@ export const deleteInterviewerTimeSlot: RequestHandler = async (
...
+    })
+  }
+}
+
+export const interviewerRankings: RequestHandler = async (
+  req: AuthRequest,
+  res
+) => {
+  try {
+    const { page, limit } = req.query
```
[the remark this answers is not in the document]
Comment (uditkumar01) [reply src/controllers/InterviewController.ts:479]: 👍

**Review by MehulKChaudhari — APPROVED** [review]
Hey Udit, Great job bro. There is a lot to learn in this PR. Checked everything locally everything is fine. Its good to go. Kudos man🔥

**Comment by MehulKChaudhari** [comment]
@NadaaFarook There is a small thing to be taken care of. Before even the interview route was open for me. The interview step had a `completed` status. Please look into this. 

gnome-shell-screenshot-N5FBA1

**Pull request merged**